lib/types.nix

```diff
@@ -17,6 +17,18 @@ rec {
     _type = typeName;
   };
 
+  # `showsPrec d t` shows the description of type `t` in an
+  # enclosing expression with precedence `d`. Since Nix does
+  # not have arbitrary infix operators, only two precedences
+  # are defined: 1 for function application, which may need
+  # to be parenthesised, and 0 for constants.
```

Comment:
We do have an infix operator for the types though: `either`. With this PR, `either int (either bool str)` shows as
```
signed integer or (boolean or string)
```

Which I think isn't correct, because `either` is left and right-associative, so it should be
```
signed integer or boolean or string
```

This would then also obviate the need for the `oneOf` change.

And I believe to implement this, we need a `descriptionPrec = d: ...` field on option types. This then also mirrors Haskell's `showsPrec` more closely.